utils.py

```
@@ -11,10 +14,286 @@
+            backup_path.unlink()
+
+    except PermissionError:
+        # Restore backup if available
+        if backup_path and backup_path.exists():
+            try:
+                import shutil
+
+                shutil.move(backup_path, path)
+            except Exception:
+                pass
+        raise FileOperationError(f"Permission denied writing to file: {path}")
```

Ruff (0.11.9)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)